File: R/example6_4.R
# Example 6.4
# Mixture of Two Normal Distributions

setwd("C:/Course17/ceR/R")
# Estimating a mixture of probability functions
# for a given set of data (sample) following a probability distribution
# find the parameters to maximize the probabity (likelihood)
#
yed<- read.table("http://web.pdx.edu/~crkl/ceR/data/yed20.txt",header=T,nrows=20)
# file name may be case sensitive!
# yed<- read.table("C:/course17/ec575/data/yed20.txt",header=T,nrows=20)
summary(yed)
y=yed$Income/10  # scale the data may help, y ~ prob. dist.

# mixture of two normal distributions
# b[1]=mu1, b[2]=sigma1, b[3]=mu2, b[4]=sigma2
llf1<-function(b) log(b[5]*dnorm(y,b[1],b[2])+(1-b[5])*dnorm(y,b[3],b[4]))
llf<-function(b) sum(llf1(b))

library(maxLik)  # need to install maxLik package
M1<-maxLik(llf,start=c(3,3,2,2,0.5))
summary(M1)

M1<-maxLik(llf1,start=c(3,3,2,2,0.5))
summary(M1)

# optim(c(3,2,2,1,0.5),llf,method="BFGS",hessian=T,control=list(fnscale=-1))
rm(list=ls())

File: python/example6_4.py
# Example 6.4
# Mixture of Two Normal Distributions

import pandas as pd 
import numpy as np
import scipy.stats as stats
from scipy import optimize

yed = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/yed20.txt', sep='\s+', nrows=20) 
y = yed.Income/10
#mixture of two normal distributions 
#b[0]=mu1, b[1]=sigma1, b[2]=mu2, b[3]=sigma2, b[4]=lambda 
def llf(b): 
    lf1 = stats.norm.pdf(y, loc=b[0], scale=b[2]) 
    lf2 = stats.norm.pdf(y, loc=b[1], scale=b[3]) 
    return -np.sum(np.log((1-b[4])*lf1 + b[4]*lf2))

b = [3, 3, 2, 2, .5] 
res1 = optimize.minimize(llf,b,method='Nelder-Mead')
b = res1.x
res2 = optimize.minimize(llf,b,method='bfgs')
print(res2)

# report and print regression output
def print_output(b,vb,bname):
    "Print Regression Output"
    se = np.sqrt(np.diag(vb))
    tr = b/se
    params = pd.DataFrame({'Parameter': b, 'Std. Error': se, 't-Ratio': tr}, index=bname)
    var_cov = pd.DataFrame(vb, index=bname, columns=bname)
    print('\nParameter Estimates')
    print(params)
    print('\nVariance-Covriance Matrix')
    print(var_cov)
    return

from numpy.linalg import inv, det
import statsmodels.tools.numdiff as nd
H = nd.approx_hess(res2.x,llf)
# check Gradient and Hessian
H  # positive definite?
det(H)
inv(H)  # not necessary the same as res['hess_inv']?
Hinv = inv(H)
var = Hinv    
# var = res2.hess_inv
print_output(res2['x'],var,['x1','x2','x3','x4','x5'])

